File: packages/dart/lib/src/tlv.dart
import 'dart:convert';
import 'dart:typed_data';

/// A single Tag-Length-Value node in the binary parse tree.
class TlvNode {
  final int offset;
  final String tag;
  final int size;
  final List<TlvNode> children;
  final Uint8List payload;

  TlvNode({
    required this.offset,
    required this.tag,
    required this.size,
    List<TlvNode>? children,
    Uint8List? payload,
  })  : children = children ?? const [],
        payload = payload ?? Uint8List(0);
}

/// Low-level TLV (Tag-Length-Value) binary parsing helpers, ported from
/// Python's _core.py. All multi-byte reads are explicitly little-endian.
class Tlv {
  /// VFF (2021+) model.dat binary structure uses Type-Length-Value (TLV) records.
  /// Most TLV tags carry raw leaf binary/string payloads, but container tags contain
  /// nested sub-TLV nodes (e.g. definitions, drawing elements, component instances).
  /// [containerTags] lists every tag hex ID that the TLV parser must recursively traverse.
  static const Set<String> containerTags = {
    'F401', 'F701', 'D430', 'D530', 'C832',
    '7C15', '8813', '8913', '8A13', '8B13', '8C13', '8D13', '4C1D', '6419',
    'F901', '7017', '7117', 'D007', 'C409', '9411', '9511', '0F01',
    '384A', 'B80B', '9713', '2C4C', 'AC0D', 'AE0D', 'F601', 'F801',
    '983A', '993A', '8C3C', '8D3C',
    // Image-entity placement: an Image placed in the model wraps a standard
    // 6419 instance node inside 9013 -> 401F. Without these two containers,
    // that inner instance stays buried in an opaque payload and the image
    // definition looks "never placed".
    '9013', '401F',
  };

  static int readU16(Uint8List data, int offset) {
    return ByteData.sublistView(data, offset, offset + 2)
        .getUint16(0, Endian.little);
  }

  static int readU32(Uint8List data, int offset) {
    return ByteData.sublistView(data, offset, offset + 4)
        .getUint32(0, Endian.little);
  }

  static int readI32(Uint8List data, int offset) {
    return ByteData.sublistView(data, offset, offset + 4)
        .getInt32(0, Endian.little);
  }

  static double readF64(Uint8List data, int offset) {
    return ByteData.sublistView(data, offset, offset + 8)
        .getFloat64(0, Endian.little);
  }

  static int parseVarInt(Uint8List data, int offset, int length) {
    int val = 0;
    for (int i = 0; i < length; i++) {
      val |= data[offset + i] << (8 * i);
    }
    return val;
  }

  static String toHexUpper(Uint8List data) {
    final buf = StringBuffer();
    for (final b in data) {
      buf.write(b.toRadixString(16).padLeft(2, '0').toUpperCase());
    }
    return buf.toString();
  }

  static String _tagHex(Uint8List data, int offset) {
    return data[offset].toRadixString(16).padLeft(2, '0').toUpperCase() +
        data[offset + 1].toRadixString(16).padLeft(2, '0').toUpperCase();
  }

  static List<TlvNode> parseRecursive(Uint8List data, int start, int end,
      [Set<String>? tags]) {
    final containerTagsSet = tags ?? containerTags;
    final elements = <TlvNode>[];
    int pos = start;
    while (pos <= end - 6) {
      final tagHex = _tagHex(data, pos);
      final size = readU32(data, pos + 2);
      if (pos + 6 + size > end) {
        break;
      }
      final isContainer = containerTagsSet.contains(tagHex);
      List<TlvNode> children = const [];
      if (isContainer && size > 0) {
        children =
            parseRecursive(data, pos + 6, pos + 6 + size, containerTagsSet);
      }
      Uint8List payload = Uint8List(0);
      if (children.isEmpty && size > 0) {
        payload = Uint8List.sublistView(data, pos + 6, pos + 6 + size);
      }
      elements.add(TlvNode(
          offset: pos,
          tag: tagHex,
          size: size,
          children: children,
          payload: payload));
      pos += 6 + size;
    }
    return elements;
  }

  /// Walk a raw payload as a flat TLV sequence (no container-tag awareness);
  /// returns (tag, body) pairs.
  static List<(String, Uint8List)> parseFlat(Uint8List payload) {
    final result = <(String, Uint8List)>[];
    int pos = 0;
    while (pos <= payload.length - 6) {
      final tag = _tagHex(payload, pos);
      final size = readU32(payload, pos + 2);
      if (pos + 6 + size > payload.length) break;
      final body = Uint8List.sublistView(payload, pos + 6, pos + 6 + size);
      result.add((tag, body));
      pos += 6 + size;
    }
    return result;
  }

  static Uint8List? findFlat(List<(String, Uint8List)> seq, String tag) {
    for (final (t, body) in seq) {
      if (t == tag) return body;
    }
    return null;
  }

  static String decodeUtf8(Uint8List bytes) {
    return utf8.decode(bytes, allowMalformed: true);
  }

  /// Scan [start, end) for direct-child (tag, offset, size) headers only,
  /// without recursing into any container - O(sibling count), not O(total
  /// node count). Used by iterTopLevelLazy to locate top-level records one
  /// at a time.
  static List<(String, int, int)> _flatHeaders(Uint8List data, int start, int end) {
    final headers = <(String, int, int)>[];
    int pos = start;
    while (pos <= end - 6) {
      final tagHex = _tagHex(data, pos);
      final size = readU32(data, pos + 2);
      if (pos + 6 + size > end) break;
      headers.add((tagHex, pos, size));
      pos += 6 + size;
    }
    return headers;
  }

  /// Yield `(index, total, node)` for each top-level TLV record's
  /// fully-recursed node one at a time, transparently unwrapping a lone
  /// "F401" wrapper - without ever materializing more than one top-level
  /// subtree simultaneously. `total` (the top-level sibling count) comes
  /// for free from the same cheap header scan that drives the loop, so
  /// callers can report "N of total" progress with no extra pass over the
  /// file. Each yielded node is safe to discard (drop all references) once
  /// the caller is done with it, before the next one is produced - that's
  /// what keeps peak memory bounded by the size of the single largest
  /// top-level record instead of the whole file (real production files can
  /// have 100k+ separate definitions).
  static Iterable<(int, int, TlvNode)> iterTopLevelLazy(Uint8List data, int start, int end, [Set<String>? tags]) sync* {
    final containerTagsSet = tags ?? containerTags;

    var headers = _flatHeaders(data, start, end);
    if (headers.length == 1 && headers[0].$1 == 'F401') {
      final (_, f401Offset, f401Size) = headers[0];
      headers = _flatHeaders(data, f401Offset + 6, f401Offset + 6 + f401Size);
    }

    final total = headers.length;
    var index = 0;
    for (final (_, offset, size) in headers) {
      final recordEnd = offset + 6 + size;
      final nodes = parseRecursive(data, offset, recordEnd, containerTagsSet);
      if (nodes.isNotEmpty) {
        yield (index, total, nodes[0]);
      }
      index++;
    }
  }
}

File: packages/dotnet/OpenSkp/Tlv.cs
using System;
using System.Collections.Generic;

namespace OpenSkp
{
    /// <summary>A single Tag-Length-Value node in the binary parse tree.
    /// Offset/Size are `long` since a decompressed model.dat can exceed
    /// 2 GB (see ChunkedBuffer) - the position within it then no longer
    /// fits in an Int32.</summary>
    internal sealed class TlvNode
    {
        public long Offset;
        public string Tag = "";
        public long Size;
        public List<TlvNode> Children = new List<TlvNode>();
        public byte[] Payload = Array.Empty<byte>();
    }

    /// <summary>Low-level TLV (Tag-Length-Value) binary parsing helpers.
    /// Ported from Python's _core.py; all multi-byte reads are explicitly
    /// little-endian regardless of host platform endianness.</summary>
    internal static class Tlv
    {
        /// <summary>VFF (2021+) model.dat binary structure uses Type-Length-Value (TLV) records.
        /// Most TLV tags carry raw leaf binary/string payloads, but container tags contain
        /// nested sub-TLV nodes (e.g. definitions, drawing elements, component instances).
        /// ContainerTags lists every tag hex ID that the TLV parser must recursively traverse.</summary>
        public static readonly HashSet<string> ContainerTags = new HashSet<string>
        {
            "F401", "F701", "D430", "D530", "C832",
            "7C15", "8813", "8913", "8A13", "8B13", "8C13", "8D13", "4C1D", "6419",
            "F901", "7017", "7117", "D007", "C409", "9411", "9511", "0F01",
            "384A", "B80B", "9713", "2C4C", "AC0D", "AE0D", "F601", "F801",
            "983A", "993A", "8C3C", "8D3C",
            // Image-entity placement: an Image placed in the model wraps a
            // standard 6419 instance node inside 9013 -> 401F. Without these
            // two containers, that inner instance stays buried in an opaque
            // payload and the image definition looks "never placed".
            "9013", "401F",
        };

        public static ushort ReadU16(byte[] data, int offset)
        {
            return (ushort)(data[offset] | (data[offset + 1] << 8));
        }

        public static uint ReadU32(byte[] data, int offset)
        {
            return (uint)(data[offset] | (data[offset + 1] << 8) | (data[offset + 2] << 16) | (data[offset + 3] << 24));
        }

        public static int ReadI32(byte[] data, int offset)
        {
            return unchecked((int)ReadU32(data, offset));
        }

        public static double ReadF64(byte[] data, int offset)
        {
            long bits = 0;
            for (int i = 0; i < 8; i++)
            {
                bits |= (long)data[offset + i] << (8 * i);
            }
            return BitConverter.Int64BitsToDouble(bits);
        }

        public static long ParseVarInt(byte[] data, int offset, int length)
        {
            long val = 0;
            for (int i = 0; i < length; i++)
            {
                val |= (long)data[offset + i] << (8 * i);
            }
            return val;
        }

        public static string ToHexUpper(byte[] data)
        {
            var chars = new char[data.Length * 2];
            for (int i = 0; i < data.Length; i++)
            {
                string h = data[i].ToString("X2");
                chars[i * 2] = h[0];
                chars[i * 2 + 1] = h[1];
            }
            return new string(chars);
        }

        private static string TagHex(byte[] data, int offset)
        {
            return data[offset].ToString("X2") + data[offset + 1].ToString("X2");
        }

        private static string TagHex(ChunkedBuffer data, long offset)
        {
            return data[offset].ToString("X2") + data[offset + 1].ToString("X2");
        }

        public static List<TlvNode> ParseRecursive(ChunkedBuffer data, long start, long end, HashSet<string>? containerTags = null)
        {
            containerTags ??= ContainerTags;
            var elements = new List<TlvNode>();
            long pos = start;
            while (pos <= end - 6)
            {
                string tagHex = TagHex(data, pos);
                uint size = data.ReadU32(pos + 2);
                if (pos + 6 + size > end)
                {
                    break;
                }
                bool isContainer = containerTags.Contains(tagHex);
                List<TlvNode> children = new List<TlvNode>();
                if (isContainer && size > 0)
                {
                    children = ParseRecursive(data, pos + 6, pos + 6 + size, containerTags);
                }
                byte[] payload = Array.Empty<byte>();
                if (children.Count == 0 && size > 0)
                {
                    payload = data.Slice(pos + 6, (int)size);
                }
                elements.Add(new TlvNode
                {
                    Offset = pos,
                    Tag = tagHex,
                    Size = size,
                    Children = children,
                    Payload = payload,
                });
                pos += 6 + size;
            }
            return elements;
        }

        /// <summary>Walk a raw payload as a flat TLV sequence (no container-tag
        /// awareness); returns [(tag, body)] pairs.</summary>
        public static List<(string Tag, byte[] Body)> ParseFlat(byte[] payload)
        {
            var result = new List<(string, byte[])>();
            int pos = 0;
            while (pos <= payload.Length - 6)
            {
                string tag = TagHex(payload, pos);
                uint size = ReadU32(payload, pos + 2);
                if (pos + 6 + size > payload.Length) break;
                byte[] body = new byte[size];
                Array.Copy(payload, pos + 6, body, 0, (int)size);
                result.Add((tag, body));
                pos += 6 + (int)size;
            }
            return result;
        }

        public static byte[]? FindFlat(List<(string Tag, byte[] Body)> seq, string tag)
        {
            foreach (var (t, body) in seq)
            {
                if (t == tag) return body;
            }
            return null;
        }

        /// <summary>Scan [start, end) for direct-child (tag, offset, size)
        /// headers only, without recursing into any container - O(sibling
        /// count), not O(total node count). Used by IterTopLevelLazy to
        /// locate top-level records one at a time.</summary>
        private static List<(string Tag, long Offset, long Size)> FlatHeaders(ChunkedBuffer data, long start, long end)
        {
            var headers = new List<(string, long, long)>();
            long pos = start;
            while (pos <= end - 6)
            {
                string tagHex = TagHex(data, pos);
                uint size = data.ReadU32(pos + 2);
                if (pos + 6 + size > end) break;
                headers.Add((tagHex, pos, size));
                pos += 6 + size;
            }
            return headers;
        }

        /// <summary>One top-level TLV record yielded by IterTopLevelLazy,
        /// with its position among the file's top-level siblings.</summary>
        public readonly struct TopLevelRecord
        {
            public readonly int Index;
            public readonly int Total;
            public readonly TlvNode Node;
            public TopLevelRecord(int index, int total, TlvNode node)
            {
                Index = index;
                Total = total;
                Node = node;
            }
        }

        /// <summary>Yield each top-level TLV record's fully-recursed node one
        /// at a time, transparently unwrapping a lone "F401" wrapper -
        /// without ever materializing more than one top-level subtree
        /// simultaneously. Total (the top-level sibling count) comes for free
        /// from the same cheap header scan that drives the loop, so callers
        /// can report "N of total" progress with no extra pass over the
        /// file. Each yielded node is safe to discard (drop all references)
        /// once the caller is done with it, before the next one is produced
        /// - that's what keeps peak memory bounded by the size of the single
        /// largest top-level record instead of the whole file (real
        /// production files can have 100k+ separate definitions).</summary>
        public static IEnumerable<TopLevelRecord> IterTopLevelLazy(ChunkedBuffer data, long start, long end, HashSet<string>? containerTags = null)
        {
            containerTags ??= ContainerTags;

            var headers = FlatHeaders(data, start, end);
            if (headers.Count == 1 && headers[0].Tag == "F401")
            {
                var (_, f401Offset, f401Size) = headers[0];
                headers = FlatHeaders(data, f401Offset + 6, f401Offset + 6 + f401Size);
            }

            int total = headers.Count;
            int index = 0;
            foreach (var (tag, offset, size) in headers)
            {
                long recordEnd = offset + 6 + size;
                var nodes = ParseRecursive(data, offset, recordEnd, containerTags);
                if (nodes.Count > 0)
                {
                    yield return new TopLevelRecord(index, total, nodes[0]);
                }
                index++;
            }
        }
    }
}
